bittensor/_neuron/text/core_validator/__init__.py

```diff
@@ -794,18 +816,26 @@ def forward(
         # Ensure number of queried neurons does not exceed metagraph.n
         num_endpoints = min([self.config.nucleus.topk, metagraph.n])
 
-        logger.info(f'Forward \t| Routing forward <dim>[{time.time() - start_time:.3g}s]</dim>')
-        logger.info(f'Dendrite \t| Request {num_endpoints} x {list(inputs_seq.shape)}')
-        request_start_time = time.time()
+        # === Ensure each UID is queried once ===
+        # Persist object variable self.permute_uids across forward calls.
+        # Reset to new permutation of all UIDs once empty.
+        if len(self.permute_uids) == 0:  # no more UIDs to query
+            self.permute_uids = torch.randperm(metagraph.n)  # reset to new permutation of all UIDs
 
         # === Randomly select num_endpoints UIDs ===
-        random_uids = torch.randperm(metagraph.n)[:num_endpoints]
+        random_uids = self.permute_uids[:num_endpoints]  # newest selection of UIDs to query
```

Comment:
self.permute_uids is a torch tensor, correct? If len(permute_uids) < num_endpoints, it wouldn't break?

Reply:
i think prev set of responsive uid was not complete, and this change makes it complete

Reply:
If `len(permute_uids) < num_endpoints` with a torch tensor it should only provide the available slice, same as with a normal list slice index that exceeds actual len. This is tested to not break during final 16 UIDs in 4096 indexed in subsets of 20, where it only slices the 16 available UIDs.